Native: Error instantiating extensions registry for FunctionSupport

Hello,

I've been having a problem for some months now on a native image. On every first report generated after my application is restarted, an exception is thrown, though the report is generated just fine and the exception isn't thrown anymore unless the app is restarted.

The exception is as follows:

```
ERROR [net.sf.jasperreports.extensions.DefaultExtensionsRegistry] Error instantiating extensions registry for functions from resource:/jasperreports_extension.properties#3: net.sf.jasperreports.engine.JRRuntimeException: java.lang.ClassNotFoundException: net.sf.jasperreports.functions.standard.ReportFunctions
	at net.sf.jasperreports.functions.FunctionsBundle.addFunctionClass(FunctionsBundle.java:88)
	at net.sf.jasperreports.functions.FunctionsBundle.<init>(FunctionsBundle.java:53)
	at net.sf.jasperreports.functions.FunctionsRegistryFactory.createRegistry(FunctionsRegistryFactory.java:57)
	at net.sf.jasperreports.extensions.DefaultExtensionsRegistry.instantiateRegistry(DefaultExtensionsRegistry.java:299)
	at net.sf.jasperreports.extensions.DefaultExtensionsRegistry.loadRegistries(DefaultExtensionsRegistry.java:274)
	at net.sf.jasperreports.extensions.DefaultExtensionsRegistry.loadRegistries(DefaultExtensionsRegistry.java:194)
	at net.sf.jasperreports.extensions.DefaultExtensionsRegistry.getRegistries(DefaultExtensionsRegistry.java:157)
	at net.sf.jasperreports.extensions.DefaultExtensionsRegistry.getExtensions(DefaultExtensionsRegistry.java:129)
	at net.sf.jasperreports.engine.DefaultJasperReportsContext.getExtensions(DefaultJasperReportsContext.java:273)
	at net.sf.jasperreports.engine.SimpleJasperReportsContext.getExtensions(SimpleJasperReportsContext.java:141)
	at net.sf.jasperreports.repo.RepositoryUtil.getServices(RepositoryUtil.java:89)
	at net.sf.jasperreports.repo.RepositoryUtil.getResourceInfo(RepositoryUtil.java:236)
	at net.sf.jasperreports.engine.JasperFillManager.getReportSource(JasperFillManager.java:1111)
	at net.sf.jasperreports.engine.JasperFillManager.fillFromRepo(JasperFillManager.java:340)
	at com.app.blahblah.Util.generateReportFile(Util.java:137)
	at com.app.blahblah.Util.generateReport(Util.java:73)
	at com.app.blahblah.Util.generateReport(Util.java:87)
	at com.app.blahblah.Bean.printData(Bean.java:2433)
	at com.app.blahblah.Bean_ClientProxy.printData(Unknown Source)
	at java.base@21.0.6/java.lang.reflect.Method.invoke(Method.java:580)
	at org.apache.el.parser.AstValue.invoke(AstValue.java:230)
	at org.apache.el.MethodExpressionImpl.invoke(MethodExpressionImpl.java:230)
	at org.apache.myfaces.view.facelets.el.ContextAwareTagMethodExpression.invoke(ContextAwareTagMethodExpression.java:96)
	at org.apache.myfaces.application.ActionListenerImpl.processAction(ActionListenerImpl.java:67)
	at org.primefaces.application.DialogActionListener.processAction(DialogActionListener.java:54)
	at jakarta.faces.component.UICommand.broadcast(UICommand.java:65)
	at jakarta.faces.component.UIViewRoot._broadcastAll(UIViewRoot.java:1248)
	at jakarta.faces.component.UIViewRoot.broadcastEvents(UIViewRoot.java:429)
	at jakarta.faces.component.UIViewRoot._process(UIViewRoot.java:1719)
	at jakarta.faces.component.UIViewRoot.processApplication(UIViewRoot.java:930)
	at org.apache.myfaces.lifecycle.InvokeApplicationExecutor.execute(InvokeApplicationExecutor.java:43)
	at org.apache.myfaces.lifecycle.LifecycleImpl.executePhase(LifecycleImpl.java:172)
	at org.apache.myfaces.lifecycle.LifecycleImpl.execute(LifecycleImpl.java:125)
	at jakarta.faces.webapp.FacesServlet.service(FacesServlet.java:223)
	at io.undertow.servlet.handlers.ServletHandler.handleRequest(ServletHandler.java:74)
	at io.undertow.servlet.handlers.security.ServletSecurityRoleHandler.handleRequest(ServletSecurityRoleHandler.java:63)
	at io.undertow.servlet.handlers.ServletChain$1.handleRequest(ServletChain.java:68)
	at io.undertow.servlet.handlers.ServletDispatchingHandler.handleRequest(ServletDispatchingHandler.java:36)
	at io.undertow.servlet.handlers.RedirectDirHandler.handleRequest(RedirectDirHandler.java:67)
	at io.undertow.servlet.handlers.security.SSLInformationAssociationHandler.handleRequest(SSLInformationAssociationHandler.java:133)
	at io.undertow.servlet.handlers.security.ServletAuthenticationCallHandler.handleRequest(ServletAuthenticationCallHandler.java:57)
	at io.undertow.server.handlers.PredicateHandler.handleRequest(PredicateHandler.java:43)
	at io.undertow.security.handlers.AbstractConfidentialityHandler.handleRequest(AbstractConfidentialityHandler.java:46)
	at io.undertow.servlet.handlers.security.ServletConfidentialityConstraintHandler.handleRequest(ServletConfidentialityConstraintHandler.java:65)
	at io.undertow.security.handlers.AuthenticationMechanismsHandler.handleRequest(AuthenticationMechanismsHandler.java:60)
	at io.undertow.servlet.handlers.security.CachedAuthenticatedSessionHandler.handleRequest(CachedAuthenticatedSessionHandler.java:77)
	at io.undertow.security.handlers.NotificationReceiverHandler.handleRequest(NotificationReceiverHandler.java:50)
	at io.undertow.security.handlers.AbstractSecurityContextAssociationHandler.handleRequest(AbstractSecurityContextAssociationHandler.java:43)
	at io.undertow.server.handlers.PredicateHandler.handleRequest(PredicateHandler.java:43)
	at io.undertow.server.handlers.PredicateHandler.handleRequest(PredicateHandler.java:43)
	at io.undertow.servlet.handlers.ServletInitialHandler.handleFirstRequest(ServletInitialHandler.java:247)
	at io.undertow.servlet.handlers.ServletInitialHandler$2.call(ServletInitialHandler.java:111)
	at io.undertow.servlet.handlers.ServletInitialHandler$2.call(ServletInitialHandler.java:108)
	at io.undertow.servlet.core.ServletRequestContextThreadSetupAction$1.call(ServletRequestContextThreadSetupAction.java:48)
	at io.undertow.servlet.core.ContextClassLoaderSetupAction$1.call(ContextClassLoaderSetupAction.java:43)
	at io.quarkus.undertow.runtime.UndertowDeploymentRecorder$8$1.call(UndertowDeploymentRecorder.java:643)
	at io.undertow.servlet.handlers.ServletInitialHandler.dispatchRequest(ServletInitialHandler.java:227)
	at io.undertow.servlet.handlers.ServletInitialHandler.handleRequest(ServletInitialHandler.java:152)
	at io.undertow.server.handlers.PathHandler.handleRequest(PathHandler.java:91)
	at io.undertow.server.handlers.CanonicalPathHandler.handleRequest(CanonicalPathHandler.java:49)
	at io.quarkus.undertow.runtime.UndertowDeploymentRecorder$1.handleRequest(UndertowDeploymentRecorder.java:126)
	at io.undertow.server.Connectors.executeRootHandler(Connectors.java:284)
	at io.undertow.server.DefaultExchangeHandler.handle(DefaultExchangeHandler.java:18)
	at io.quarkus.undertow.runtime.UndertowDeploymentRecorder$4$2.run(UndertowDeploymentRecorder.java:443)
	at java.base@21.0.6/java.util.concurrent.Executors$RunnableAdapter.call(Executors.java:572)
	at java.base@21.0.6/java.util.concurrent.FutureTask.run(FutureTask.java:317)
	at io.quarkus.vertx.core.runtime.VertxCoreRecorder$15.runWith(VertxCoreRecorder.java:639)
	at org.jboss.threads.EnhancedQueueExecutor$Task.doRunWith(EnhancedQueueExecutor.java:2675)
	at org.jboss.threads.EnhancedQueueExecutor$Task.run(EnhancedQueueExecutor.java:2654)
	at org.jboss.threads.EnhancedQueueExecutor.runThreadBody(EnhancedQueueExecutor.java:1627)
	at org.jboss.threads.EnhancedQueueExecutor$ThreadBody.run(EnhancedQueueExecutor.java:1594)
	at org.jboss.threads.DelegatingRunnable.run(DelegatingRunnable.java:11)
	at org.jboss.threads.ThreadLocalResettingRunnable.run(ThreadLocalResettingRunnable.java:11)
	at io.netty.util.concurrent.FastThreadLocalRunnable.run(FastThreadLocalRunnable.java:30)
	at java.base@21.0.6/java.lang.Thread.runWith(Thread.java:1596)
	at java.base@21.0.6/java.lang.Thread.run(Thread.java:1583)
	at org.graalvm.nativeimage.builder/com.oracle.svm.core.thread.PlatformThreads.threadStartRoutine(PlatformThreads.java:896)
	at org.graalvm.nativeimage.builder/com.oracle.svm.core.thread.PlatformThreads.threadStartRoutine(PlatformThreads.java:872)
Caused by: java.lang.ClassNotFoundException: net.sf.jasperreports.functions.standard.ReportFunctions
	at org.graalvm.nativeimage.builder/com.oracle.svm.core.hub.ClassForNameSupport.forName(ClassForNameSupport.java:122)
	at org.graalvm.nativeimage.builder/com.oracle.svm.core.hub.ClassForNameSupport.forName(ClassForNameSupport.java:86)
	at java.base@21.0.6/java.lang.Class.forName(DynamicHub.java:1359)
	at java.base@21.0.6/java.lang.Class.forName(DynamicHub.java:1348)
	at net.sf.jasperreports.engine.util.JRClassLoader.loadClassForRealName(JRClassLoader.java:208)
	at net.sf.jasperreports.engine.util.JRClassLoader.loadClassForName(JRClassLoader.java:148)
	at net.sf.jasperreports.functions.FunctionsBundle.addFunctionClass(FunctionsBundle.java:83)
	... 77 more
```

Looks like this is a case for registering some classes for reflection. I tried registering the mentioned class, ```ReportFunctions.class```, but other classes showed up. The exception stopped appearing after 5 classes were registered for reflection. These are:

```
DateTimeFunctions.class,
LogicalFunctions.class,
MathFunctions.class,
ReportFunctions.class,
TextFunctions.class
```

I hope this helps. I can't remember when exactly this started happening, but it's been present in my application for months, if not more than a year...

--- a/deployment/src/main/java/io/quarkiverse/jasperreports/deployment/AbstractJandexProcessor.java
+++ b/deployment/src/main/java/io/quarkiverse/jasperreports/deployment/AbstractJandexProcessor.java
@@ -70,7 +70,7 @@ abstract class AbstractJandexProcessor {
             subclasses.addAll(collectSubclasses(combinedIndex, implementationClass));
         }
         classes.addAll(subclasses);
-        Log.debugf("Implementors: %s", classes);
+        Log.debugf("%S: Implementors: %s", className, classes);
         return new ArrayList<>(classes);
     }
 }

--- a/deployment/src/main/java/io/quarkiverse/jasperreports/deployment/JasperReportsProcessor.java
+++ b/deployment/src/main/java/io/quarkiverse/jasperreports/deployment/JasperReportsProcessor.java
@@ -151,6 +151,7 @@ class JasperReportsProcessor extends AbstractJandexProcessor {
         //@formatter:off
         final List<String> classNames = new ArrayList<>();
         // By Implementors: jasper interfaces/abstract classes that are created with Class.forName
+        classNames.addAll(collectImplementors(combinedIndex, net.sf.jasperreports.functions.FunctionSupport.class.getName()));
         classNames.addAll(collectImplementors(combinedIndex, net.sf.jasperreports.engine.JRDataSource.class.getName()));
         classNames.addAll(collectImplementors(combinedIndex, net.sf.jasperreports.engine.JRDataSourceProvider.class.getName()));
         classNames.addAll(collectImplementors(combinedIndex, net.sf.jasperreports.engine.JRTemplate.class.getName()));
@@ -162,8 +163,8 @@ class JasperReportsProcessor extends AbstractJandexProcessor {
         classNames.addAll(collectImplementors(combinedIndex, net.sf.jasperreports.engine.xml.ReportLoader.class.getName()));
         classNames.addAll(collectImplementors(combinedIndex, net.sf.jasperreports.extensions.ExtensionsRegistry.class.getName()));
         classNames.addAll(collectImplementors(combinedIndex, net.sf.jasperreports.extensions.ExtensionsRegistryFactory.class.getName()));
-        classNames.addAll(collectSubclasses(combinedIndex, net.sf.jasperreports.engine.JRAbstractExporter.class.getName()));
         classNames.addAll(collectSubclasses(combinedIndex, net.sf.jasperreports.dataadapters.AbstractDataAdapter.class.getName()));
+        classNames.addAll(collectSubclasses(combinedIndex, net.sf.jasperreports.engine.JRAbstractExporter.class.getName()));
 
         // By Package (utilities etc)
         classNames.addAll(collectClassesInPackage(combinedIndex, net.sf.jasperreports.compilers.ReportExpressionEvaluationData.class.getPackageName()));
@@ -183,6 +184,7 @@ class JasperReportsProcessor extends AbstractJandexProcessor {
         classNames.addAll(collectClassesInPackage(combinedIndex, net.sf.jasperreports.engine.xml.ReportLoader.class.getPackageName()));
         classNames.addAll(collectClassesInPackage(combinedIndex, net.sf.jasperreports.export.CsvExporterConfiguration.class.getPackageName()));
         classNames.addAll(collectClassesInPackage(combinedIndex, net.sf.jasperreports.extensions.DefaultExtensionsRegistryFactory.class.getPackageName()));
+        classNames.addAll(collectClassesInPackage(combinedIndex, net.sf.jasperreports.functions.standard.MathFunctions.class.getPackageName()));
         classNames.addAll(collectClassesInPackage(combinedIndex, net.sf.jasperreports.governors.GovernorExtensionsRegistryFactory.class.getPackageName()));
         classNames.addAll(collectClassesInPackage(combinedIndex, net.sf.jasperreports.jackson.util.JacksonUtil.class.getPackageName()));
         classNames.addAll(collectClassesInPackage(combinedIndex, net.sf.jasperreports.parts.PartComponentsExtensionsRegistryFactory.class.getPackageName()));
@@ -271,7 +273,6 @@ class JasperReportsProcessor extends AbstractJandexProcessor {
             BuildProducer<RuntimeInitializedPackageBuildItem> runtimeInitializedPackages) {
         //@formatter:off
         List<String> classes = collectImplementors(combinedIndex, net.sf.jasperreports.extensions.ExtensionsRegistryFactory.class.getName());
-        classes.addAll(collectImplementors(combinedIndex,  net.sf.jasperreports.functions.FunctionSupport.class.getName()));
         classes.addAll(Stream.of(
                 "javax.swing",
                 "javax.swing.plaf.metal",

